### PR TITLE
flatpak-autoinstall: install the new Hack 2 Clubhouse

### PR DESCRIPTION
**Pull request by erikos**
Remove the old Hack 1 Clubhouse

https://phabricator.endlessm.com/T28277

### CODE REVIEW

File: data/50-default.json [hunk 1]
```diff
@@ -85,5 +85,29 @@
...
+  {
+    "action": "uninstall",
+    "serial": 2019102200,
+    "ref-kind": "app",
+    "collection-id": "com.endlessm.Apps.Hack",
+    "remote": "hack-apps",
+    "name": "com.endlessm.Clubhouse",
+    "branch": "stable",
+    "filters": {
+        "~architecture": ["armhf"]
```
Comment (wjt) [inline data/50-default.json]: I don't think the arch filter is needed here. We want the old Clubhouse to be removed on any system it is installed on. It just so happens that we only ever shipped it on x86_64.

File: data/50-default.json [hunk 2]
```diff
@@ -85,5 +85,29 @@
...
+  },
+  {
+    "action": "uninstall",
+    "serial": 2019102200,
+    "ref-kind": "app",
+    "collection-id": "com.endlessm.Apps.Hack",
+    "remote": "hack-apps",
+    "name": "com.endlessm.Clubhouse",
+    "branch": "stable",
+    "filters": {
```
Comment (wjt) [inline data/50-default.json:109]: Again – I think we should just drop the filter here. We don't want `com.endlessm.Clubhouse` to be installed anywhere.

File: data/50-default.json [hunk 3]
```diff
@@ -85,5 +85,26 @@
...
     "branch": "stable"
+  },
+  {
+    "action": "install",
+    "serial": 2019102100,
+    "ref-kind": "app",
+    "collection-id": "com.endlessm.Apps",
+    "remote": "eos-apps",
+    "name": "com.hack_computer.Clubhouse",
+    "branch": "stable",
```
Comment (wjt) [inline data/50-default.json]: Is this the correct branch? I don't think so – the default branch in `eos-apps` is `eos3`. Did you test that this file works? You can copy it into `/usr/share/eos-application-tools/flatpak-autoinstall.d/` and run `eos-updater-flatpak-installer --dry-run`.

Comment (erikos) [reply data/50-default.json]: We have been using master/stable for Hack apps so far. We can move that to eos3 now that the Clubhouse is part of eos-apps. I am fine with that.

Thanks for the testing suggestion. I have tried with the master branch for now, but the filter does not work. With the current filter the installation is not scheduled. If I negotiate the amd64 rule it is scheduled.

It works to negotiate both arm archs: 
`"~architecture": ["arm64", "armhf"]`

This is the Asus device with a master image. The OSTree refspec says "eos:os/eos/amd64/master".

Comment (wjt) [reply data/50-default.json]: I think Flatpak calls the arch `x86_64`, sorry.

Comment (erikos) [reply data/50-default.json]: Ahh, I compared with the archs in eos-image-builder. Ok that works now.

What about the branch, shall I keep stable? Or switch to eos3? I do not feel strongly at all.

Comment (wjt) [reply data/50-default.json]: The app should be released as `eos3` like all other apps in `eos-apps`. We don't typically vary the branch between the dev / demo / prod repos.

Comment (erikos) [reply data/50-default.json]: Ok, changed to use the eos3 branch for the Clubhouse.

**Review by wjt — COMMENTED** [review]
@pwithnall am I right in thinking that if one updates to an ostree containing this change while `com.hack_computer.Clubhouse//eos3` is not present in eos-apps, then it will not be installed, even when you install a future OS update and the app has appeared in `eos-apps`?

The reason I ask is twofold:

* If we merge this PR, my computer will be in this state tomorrow; but more importantly,
* When we release 3.7.2~beta1, we will presumably not also have released the Clubhouse (?) and so beta users will also hit this.

A workaround would be to bump the serial after publishing the app but before releasing 3.7.2 final.

**Comment by ramcq** [comment]
> @pwithnall am I right in thinking that if one updates to an ostree containing this change while `com.hack_computer.Clubhouse//eos3` is not present in eos-apps, then it will not be installed, even when you install a future OS update and the app has appeared in `eos-apps`?
> 
> The reason I ask is twofold:
> 
>     * If we merge this PR, my computer will be in this state tomorrow; but more importantly,
> 
>     * When we release 3.7.2~beta1, we will presumably not also have released the Clubhouse (?) and so beta users will also hit this.
> 
> 
> A workaround would be to bump the serial after publishing the app but before releasing 3.7.2 final.

Or don't cherry pick this from master until we're ready to release the clubhouse (and OS) publicly. So we would have to go out of our way to QA this on master.

File: data/50-default.json [hunk 4]
```diff
@@ -85,5 +85,26 @@
...
+    "branch": "eos3",
+    "filters": {
+        "architecture": ["x86_64"]
+    }
+  },
+  {
+    "action": "uninstall",
+    "serial": 2019102200,
+    "ref-kind": "app",
+    "collection-id": "com.endlessm.Apps.Hack",
```
Comment (ramcq) [inline data/50-default.json]: uninstall actions don't need collection IDs or remotes, they match the local app ID

**Comment by erikos** [comment]
Note: the quests for the OS rely on the System app to be present. We should install com.hack_computer.OperatingSystemApp as well. Is around 14 MB

**Comment by pwithnall** [comment]
> @pwithnall am I right in thinking that if one updates to an ostree containing this change while `com.hack_computer.Clubhouse//eos3` is not present in eos-apps, then it will not be installed, even when you install a future OS update and the app has appeared in `eos-apps`?

Nope, the eos-updater ‘fetch’ operation will fail before the OS update has even been applied. See the `prepare_flatpaks_to_deploy()` function in eos-updater.

Also see `eufi_apply_flatpak_ref_actions()` in eos-updater, which shows that if an OS update is (somehow) successfully downloaded and deployed, on reboot `eos-updater-flatpak-installer` will apply all the actions in the JSON file up to the one with the missing ref, and then stop there. On the next reboot, it will try the failing action again. If that succeeds, it’ll then continue to try and apply all the following actions in the JSON.

File: data/50-default.json (hunk 1 above)
Comment (erikos) [reply data/50-default.json]: Yes, removed.

File: data/50-default.json (hunk 4 above)
Comment (erikos) [reply data/50-default.json]: Good catch, resolved.

**Comment by erikos** [comment]
This is ready for review again. Now that we have green light for deploying Hack 2 with an OTA upgrade to all Endless users we are good to go.

The eos3 branches have been created.

I added as well the System app and the ProjectLibrary for the minimal Hack 2 experience.

File: data/50-default.json [hunk 5]
```diff
@@ -85,5 +85,48 @@
...
+        "architecture": ["x86_64"]
+    }
+  },
+  {
+    "action": "install",
+    "serial": 2019102300,
+    "ref-kind": "app",
+    "collection-id": "com.endlessm.Apps",
+    "remote": "eos-apps",
+    "name": "com.hack_computer.ProjectLibrary",
```
Comment (wjt) [inline data/50-default.json:119]: Have you tested whether this causes `com.hack_computer.ProjectLibrary.Content` to be installed too? The reason I ask is that I installed the clubhouse from the app center, and installed these other two from the command line, and I noticed that com.hack_computer.Clippy.Extension and com.hack_computer.ProjectLibrary.Content are pulled in too. I'd expect that eos-updater-flatpak-installer would behave the same way but I'm not sure.

**Review by wjt — APPROVED** [review]

**Pull request merged**